File: Java/Hard/Accepted/find-minimum-time-to-finish-all-jobs_2.java
class Solution {
    public int minimumTimeRequired(int[] jobs, int k) {
        
        int n = jobs.length;
        int[] sum = new int[1 << n];
        for (int i = 1; i < (1 << n); i++) {
            int x = Integer.lowestOneBit(i);
            sum[i] = sum[i - x] + jobs[Integer.numberOfTrailingZeros(x)];
        }
        int[][] dp = new int[k][1 << n];
        for (int i = 0; i < (1 << n); i++) {
            dp[0][i] = sum[i];
        }
        for (int i = 1; i < k; i++) {
            for (int j = 0; j < (1 << n); j++) {
                int min = Integer.MAX_VALUE;
                for (int x = j; x > 0; x = (x - 1) & j) {
                    min = Math.min(min, Math.max(dp[i - 1][j - x], sum[x]));
                }
                dp[i][j] = min;
            }
        }
        return dp[k - 1][(1 << n) - 1];
    }
}

File: Rust/Hard/CompileError/find-minimum-time-to-finish-all-jobs_2.rs
impl Solution {
    pub fn minimum_time_required(jobs: Vec<i32>, k: i32) -> i32 {
        
        let mut jobs = jobs;
        jobs.sort();
        let k = k as usize;
        let mut ans = std::i32::MAX;
        let mut workers = vec![0; k];
        let mut used = vec![false; k];
        let mut dfs = |jobs: &mut Vec<i32>, workers: &mut Vec<i32>, used: &mut Vec<bool>, idx: usize, max: &mut i32| {
            if idx == jobs.len() {
                *max = std::cmp::min(*max, workers.iter().max().unwrap().clone());
                return;
            }
            let mut seen = std::collections::HashSet::new();
            for i in 0..k {
                if used[i] || seen.contains(&workers[i]) {
                    continue;
                }
                seen.insert(workers[i]);
                used[i] = true;
                workers[i] += jobs[idx];
                dfs(jobs, workers, used, idx + 1, max);
                workers[i] -= jobs[idx];
                used[i] = false;
            }
        };
        dfs(&mut jobs, &mut workers, &mut used, 0, &mut ans);
        ans
    }
}